Set `ReadAdvice#NORMAL` on files that have a forward-only access pattern.

jpountz opened the pull request:
This applies to files where performing readahead could help:
 - Doc values data (`.dvd`)
 - Norms data (`.nvd`)
 - Docs and freqs in postings lists (`.doc`)
 - Points data (`.kdd`)

Other files (KNN vectors, stored fields, term vectors) keep using a `RANDOM`
advice.

jpountz committed:
Give a hint to `IndexInput` about slices that have a forward-only access pattern.

This introduces a new API that allows directories to optimize access to
`IndexInput`s that have a forward-only access pattern by reading ahead of the
current position. It would be applicable to:
 - Postings lists,
 - Doc values,
 - Norms,
 - Points.

Relates #13179

jpountz committed:
iter

rmuir commented on lucene/core/src/java21/org/apache/lucene/store/MemorySegmentIndexInput.java:
```diff
@@ -370,6 +370,16 @@ public void prefetch(long offset, long length) throws IOException {
+    }
+    // TODO: Is there a hint we can give to the OS to let it optimize for our forward-only access
+    // pattern in the given range?
+  }
```
`MADV_SEQUENTIAL`

rmuir replied:
or even better, `MADV_NORMAL` for the range, which is less aggressive.

jpountz replied:
I thought you had concerns about `MADV_NORMAL` but I like it better too, I updated the PR to use `ReadAdvice.NORMAL` instead.

jpountz committed:
Revert "iter"

This reverts commit 8375caf96284853877be544ec8366854e226a447.

jpountz committed:
Revert "Give a hint to `IndexInput` about slices that have a forward-only access pattern."

This reverts commit 8a5cb8fa1890c4bba3c0139c968eebf725e009e3.

jpountz committed:
Use `ReadAdvice#NORMAL` on files that have a forward-only access pattern.

This applies to files where performing readahead could help:
 - Doc values data (`.dvd`)
 - Norms data (`.nvd`)
 - Docs and freqs in postings lists (`.doc`)
 - Points data (`.kdd`)

Other files (KNN vectors, stored fields, term vectors) keep using a `RANDOM`
advice.
